src/core/or/relay.c

```diff
@@ -579,11 +581,20 @@ relay_send_command_from_edge_,(streamid_t stream_id, circuit_t *circ,
 
   /* Add random bytes to the unused portion of the payload, to foil attacks
    * where the other side can predict all of the bytes in the payload and thus
-   * compute authenticated sendme cells without seeing the traffic.  See
-   * proposal 289. */
+   * compute authenticated sendme cells without seeing the traffic. See
+   * proposal 289.
+   *
+   * We'll skip the first 4 bytes of unused data because having some unused
+   * zero bytes has saved us a lot of times in the past. */
+  random_bytes_len = RELAY_PAYLOAD_SIZE -
```

Comment:
I'm a little scared by signed/unsigned things when we're checking for negative numbers.  Can we have a test for this?  It might help to put it in a separate function.  Or maybe it would be easier to wrap this part of the code in something like:
```
offset = RELAY_HEADER_SIZE + PAYLOAD_LEN + 4
if (offset < RELAY_PAYLOAD_SIZE) {
   len = RELAY_PAYLOAD_SIZE - offset;
   // ...
}
```

Reply:
Sooooo... good thing you requested that because the code was actually not correct. The offset should have been tested against `CELL_PAYLOAD_SIZE` since we offset on `cell.payload`.

I've pushed a new commit that creates two helper function for the random padding and adds a unit test to test the offset value we use. I would recommend reviewing carefully here because it is very confusing to have `payload_len` not be the length of `cell.payload` but rather the *data* length.

See commit: 09f67b45e1f333a1d018d78d5311b0d54a8f2e7d